Implement datapath trace notification aggregation

joestringer opened the pull request:
This PR broadly changes two aspects of Cilium:
* Rework the configuration for the daemon and endpoints to allow integer-backed mutable configuration at runtime
* Implement a new mutable option, `MonitorAggregationLevel` which ratelimits trace notifications from the datapath up to monitor processes.

The details of the monitor aggregation levels are held in `pkg/option/monitor.go`, excerpt below:

```
        // MonitorAggregationLevelNone represents no aggregation in the         
        // datapath; all packets will be monitored.                             
        MonitorAggregationLevelNone = 0                                         
                                                                                
        // MonitorAggregationLevelLow represents aggregation of monitor events  
        // to emit a maximum of one trace event per packet. Trace events when   
        // packets are received are disabled.                                   
        MonitorAggregationLevelLowest = 1
```

Internally these are stored as integers, but in user interactions they may be configured using integers or convenient strings such as "none", "lowest", "low", "medium", "maximum", etc.

Currently only the lowest aggregation is implemented, that is to hide "from" traces, ie trace observation points when Cilium initially sees a packet. 

To break down the current maximum level of aggregation (low), it will always send:
* All debug notifications if `Debug` is `Enabled`.
* All drop notifications if `DropNotification` is `Enabled`.
* Trace messages:
  * If `TraceNotification` option is `Enabled`, notifications for `TRACE_TO_XXX` observation points.

It will not: 
* Send notifications for `TRACE_FROM_XXX` observation points.

The following planned aggregation modes are not yet implemented:
* Medium: Only send traces periodically for active connections, or when there is new information such as tcp flags changes (~Proposed~ Merged in #4936)
* High: Also ratelimit the drop monitor messages

Related: #4080

Tasks:
* [x] Linux 4.9 is reporting complexity explosion (>65K instructions)
  * Related to Medium aggregation mode, pushed out to another PR
* [x] Atomic access of ct entry flags/timestamps
  * Related to Medium aggregation mode, pushed out to another PR
* [x] Restore of endpoints created on an older version may fail to initialize the `MonitorAggregationLevel` config option, so the option cannot be reconfigured until the daemon is restarted again.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4828)
<!-- Reviewable:end -->

joestringer committed:
examples: Add MONITOR_AGGREGATION_LEVEL option

Add the MONITOR_AGGREGATION_LEVEL option to the kubernetes YAMLs, to
allow users to specify how much aggregation should occur in the datapath
for monitor output.

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm reviewed (commented):
LGTM except the API breakage for bool options that we have discussed on slack

ianvernon commented on pkg/option/option.go:
```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
+// SetValidated sets the option `key` to the specified value. The caller is
+// expected to have validated the input to this function.
+func (io *IntOptions) SetValidated(key string, value int) {
+	io.optsMU.Lock()
```
Why not have this validate the key and value regardless? There can be a private function which can assume that validation is done which can be called internally within `pkg/option`. I dislike assuming that the caller has done anything, as it means the code is more prone to misuse.

joestringer replied:
Because sometimes you have pre-validated values and there's no point in validating things twice. For example, this function is called from the Endpoint package where the values are taken directly from the default definitions from the option package. 

Most of the usages of `Set()`, I've converted over to the new `SetBool()`, then I've renamed `Set()` to `SetValidated()` to make this more explicit, so the developer really can't ignore the fact that they're passing a value to the function that assumes the input was validated. I think that it's fine to assume particular behaviour from a caller so long as you make it very explicit, and this change makes it more explicit than it was previously.

We can always make more improvements to this, but I'll focus my effort on getting the functionality of this PR working first.

ianvernon commented on pkg/option/option.go:
```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
-	bo.optsMU.Lock()
-	if _, exists := bo.Opts[key]; !exists {
-		bo.Opts[key] = value
+func (io *IntOptions) SetIfUnset(key string, value int) {
```
This is extremely nit-picky, but I dislike the use of 'io' as a receiver name because it may look like we are using https://golang.org/pkg/io/ instead of int options.

ianvernon commented on pkg/option/option.go:
```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
-		if !enabled {
-			return "", false, fmt.Errorf("Invalid boolean format")
+		if result == OptionDisabled {
+			return "", OptionDisabled, fmt.Errorf("Invalid boolean format")
```
Is this error message still valid given that we use integers ?

joestringer replied:
Yes. It's still valid to specify an option value as something like `!debug` and if you've instead specified `!debug=true`, it makes no sense and should be rejected as an invalid boolean option.

ianvernon commented on pkg/option/option.go:
```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
+	return "", OptionDisabled, fmt.Errorf("Invalid option format")
+}
+
+func ParseKeyValue(lib *OptionLibrary, arg, value string, defaultValue int) (string, int, error) {
```
Unit tests?

joestringer replied:
Filed #4866

ianvernon commented on pkg/option/option.go:
```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
-	return bo.isEnabled(key)
+// SetValidated sets the option `key` to the specified value. The caller is
+// expected to have validated the input to this function.
+func (io *IntOptions) SetValidated(key string, value int) {
```
Unit tests?

joestringer replied:
Filed #4866

ianvernon commented on pkg/option/option.go:
```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
-func (bo *BoolOptions) isEnabled(key string) bool {
-	set, exists := bo.Opts[key]
-	return exists && set
+func (io *IntOptions) GetValue(key string) int {
```
Unit tests?

joestringer replied:
Filed #4866

ianvernon commented on pkg/option/monitor.go:
```diff
@@ -0,0 +1,114 @@
+
+// VerifyMonitorAggregationLevel validates the specified key/value for a
+// monitor aggregation level.
+func VerifyMonitorAggregationLevel(key, value string) error {
```
Unit tests?

ianvernon commented on pkg/option/monitor.go:
```diff
@@ -0,0 +1,114 @@
+// ParseMonitorAggregationLevel turns a string into a monitor aggregation
+// level. The string may contain an integer value or a string representation of
+// a particular monitor aggregation level.
+func ParseMonitorAggregationLevel(value string) (int, error) {
```
Unit tests?

ianvernon reviewed (dismissed):
I like how we are moving to a mode to allow for more rich configuration that can be applied to the datapath as opposed to just two values. I'd personally prefer if this was split out into two separate PRs, as the BoolOption --> IntOption one can be an enhancement which should not change existing behavior. Then, as a second PR, put the trace aggregation stuff on top of it. I also would like more unit tests to be added, as there are new functions added which are unit-testable.

joestringer committed:
endpoint: Rename Opts -> Options

Previously, the BoolOptions for the endpoint were stored in the `Opts`
field. This would be serialized to the filesystem during endpoint
regeneration, which could be used during restore.

An upcoming commit will instead represent the options as IntOptions. The
underlying type is different (map[string]bool becomes map[string]int),
so it causes upgrade problems if the field is named the same.

To work around this issue, rename the field from `Opts` -> `Options`.
Subsequent patches should provide proper upgrade/downgrade guardrails to
handle the conversion between these fields properly. Github issue #4865
tracks this work.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
option: Refactor BoolOptions into IntOptions

Refactor the existing boolean operations to instead allow integer
representations. This reworks the internal storage of the option from a
true/false boolean to instead use an integer, with OptionDisabled (0)
and OptionEnabled (1) representing the existing options.

The logic for checking the values for these int options is reversed to
default to checking for 0 value (disabled) first, then treating all
non-zero values as enabled. Future multiple-value options will require
additional function definitions to control the handling of non-zero
values.

The following commits will be the first user of the integer
representation, allowing a single configuration option to be configured
with multiple values.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Add MonitorAggregation option

Add the first level of monitor aggregation, which reduces the tracing of
packets to once per packet by deferring trace notifications until the
packet is transmitted from Cilium to something outside of Cilium. In
other words, the trace points for when a packet is initially received
from Cilium are disabled.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon reviewed (approved):
Approved since we agreed this can be merged in its current state; follow-ups I suggested will be taken care of separately.
